docs/metrics-data-model.md

```diff
@@ -0,0 +1,177 @@
+* In the case the two types (`Sample` and `TSample`) are confirmed, we need the `TagSet` in the Sample or can we just simplify it as a `[]string` or `map[string]string` and delegate it in a centralized place the `TagSet` association?
+* Currently, the metric type is defined by the `metrics.Metric` but the same type handle Thresholds and Sink operation. Should the `TimeSeries` type depend on `Metric`? Or should we wait to split them in a different struct?
+
+#### Tags (aka [k6/1831](https://github.com/grafana/k6/issues/1831))
+
+* Tag{Key, Value}
+* TagSet ([]*Tag)
+
+#### Sink
+
+The sinks are implemented by metric types and they keep the series values up to date:
+
+* Counter:  a monotonic increasing value
+* Gauge: the latest value
+* SparseHistogram: counters per dynamic ranges (buckets) - https://grafana.com/blog/2021/11/03/how-sparse-histograms-can-improve-efficiency-precision-and-mergeability-in-prometheus-tsdb
+
+#### Series IDs
+
+The series IDs are a `uint64` value generated by hashing the metric name and the tags' key-value pairs.
+
+#### Timestamp
+
+The timestamp is a `uint64` value representing a Unix nano.
+
+## Storage
+
+#### Time Series database (aka tsdb)
+
+We need to store all the time series generated from `k6` run during the execution so the other components (mostly the outputs) could query the storage for getting the value of the time series. It's expected to be in-memory and it should be concurrent-safe.
+
+```go
+type Repository interface {
+    InsertSeries(TimeSeries) error
+    GetSeriesByID(hash uint64) (*TimeSeries, error) 
+    GetSeries() ([]*TimeSeries, error) // maybe, we should try to avoid it
+}
+```
+
+#### TagMap
+
+```go
+type TagMap struct {
+    m sync.Map // potentially something better
+}
+```
+
+It creates a finite set of tags storing the pointer of Tags by its hash. In this way, it should reduce the used memory and the GC operations.
+
+#### Open questions:
+
+* Is the same structure by TagSet required? It could be useful in the case we need to compare a pair of TagSet, where we can just execute single access by hash instead of N comparisons.
+* Should it be possible for any reason to fetch raw data points (not aggregated)? Like `samples := tsdb.GetSamplesBySeries(seriesID, timeStart, timeEnd)`? 
+
+## Collect
+
+#### Open questions:
+
+* Should we change something in `metrics.PushIfNotDone()`?
+* Do we require to identify the time series before invoking `metrics.PushIfNotDone`? So, does the sample require to reference the time series?
+
+## metrics.Ingester
+
+TBD
+
+## Outputs
+
+Outputs should interact with the `tsdb` for fetching a time series snapshot. `series := tsdb.GetSeries()`.
```

Comment:
Again - some outputs will require more than an aggregated snapshot, they might require access to the raw samples.